lib/Backend/LowerMDShared.cpp

```diff
@@ -9307,15 +9324,52 @@ void LowererMD::GenerateFastInlineBuiltInCall(IR::Instr* instr, IR::JnHelperMeth
 
                 //MOVSD dst, src1;
                 Assert(!dst->IsEqual(src1));
-                this->m_lowerer->InsertMove(dst, src1, instr);
+                
+                IR::Opnd * opndNaN = nullptr;
+
+                bool isF32 = dst->IsFloat32();
+
+                if (isF32)
+                {
+                    opndNaN = IR::MemRefOpnd::New((double*)&(Js::JavascriptNumber::k_Nan32), IRType::TyFloat32, this->m_func);
+                }
+                else
+                {
+                    opndNaN = IR::MemRefOpnd::New((double*)&(Js::JavascriptNumber::k_Nan), IRType::TyFloat64, this->m_func);
+                }
+
+                IR::Opnd * src1f64 = src1;
+                IR::Opnd * src2f64 = src2; 
+                if (propagateNaN)
+                {
+                    if (src1->IsFloat32())
+                    {
+                        src1f64 = IR::RegOpnd::New(TyFloat64, m_func);
+                        EmitFloat32ToFloat64(src1f64, src1, instr);
```

Comment:
The result is ultimately still f32. The f64 is for getting the same comparisons for both f32 and f64. I could alternatively have f32 and f64-specific paths but that has so far turned out to be messy.